# Customer Segmentation and CLTV Prediction Project Overview

## Introduction
In today's competitive business environment, understanding customer behavior and maximizing customer lifetime value (CLTV) are crucial for companies. Customer segmentation and CLTV prediction are two powerful techniques used by businesses to gain insights into customer habits and optimize marketing strategies.

## Business Problem
FLO, one of the largest shoe companies in Turkey, aims to enhance its marketing strategies and improve customer relationship management. To achieve this, FLO plans to conduct two analyses:

1. **Customer Segmentation with RFM Analysis:** Segmenting customers based on their recency, frequency, and monetary (RFM) values to better understand their purchasing habits and tailor marketing efforts accordingly.

2. **CLTV Prediction with BG-NBD and Gamma-Gamma:** Predicting the future CLTV of customers using the Beta-Geometric/Negative Binomial Distribution (BG-NBD) model and the Gamma-Gamma submodel to identify high-value customers and allocate resources effectively.

## Dataset Overview
The dataset used for both analyses contains information on customers' past shopping behaviors with FLO as OmniChannel shoppers (both online and offline) between the years 2020 and 2021. Here is an overview of the dataset:

### Metadata:
- **master_id:** Unique customer ID
- **order_channel:** Channel used for shopping (Android, iOS, Desktop, Mobile)
- **last_order_channel:** The channel where the most recent purchase was made
- **first_order_date:** Date of the customer's first purchase
- **last_order_date:** Date of the customer's last purchase
- **last_order_date_online:** Date of the last purchase made by the customer on the online platform
- **last_order_date_offline:** Date of the last purchase made by the customer on the offline platform
- **order_num_total_ever_online:** Total number of purchases made by the customer on the online platform
- **order_num_total_ever_offline:** Total number of purchases made by the customer on the offline platform
- **customer_value_total_ever_online:** Total payment by the customer for their online shopping
- **customer_value_total_ever_offline:** Total payment by the customer for their offline shopping
- **interested_in_categories_12:** List of categories the customer has shopped in the last 12 months

### Prepare the Dataset

In [1]:
import pandas as pd
import datetime as dt


df = pd.read_csv("C:\\Users\\utku\\Desktop\\miuul\\FLOMusteriSegmentasyonu\\flo_data_20k.csv")

In [2]:
df.head(10)



master_id order_channel last_order_channel  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f   Android App            Offline   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f   Android App             Mobile   
2  69b69676-1a40-11ea-941b-000d3a38a36f   Android App        Android App   
3  1854e56c-491f-11eb-806e-000d3a38a36f   Android App        Android App   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       Desktop            Desktop   
5  e585280e-aae1-11e9-a2fc-000d3a38a36f       Desktop            Offline   
6  c445e4ee-6242-11ea-9d1a-000d3a38a36f   Android App        Android App   
7  3f1b4dc8-8a7d-11ea-8ec0-000d3a38a36f        Mobile            Offline   
8  cfbda69e-5b4f-11ea-aca7-000d3a38a36f   Android App        Android App   
9  1143f032-440d-11ea-8b43-000d3a38a36f        Mobile             Mobile   

  first_order_date last_order_date last_order_date_online  \
0       2020-10-30      2021-02-26             2021-02-21   
1       2017-02-08      2021-02-16             2021-02-16   
2       2019-11-27      2020-11-27             2020-11-27   
3       2021-01-06      2021-01-17             2021-01-17   
4       2019-08-03      2021-03-07             2021-03-07   
5       2018-11-18      2021-03-13             2018-11-18   
6       2020-03-04      2020-10-18             2020-10-18   
7       2020-05-15      2020-08-12             2020-05-15   
8       2020-01-23      2021-03-07             2021-03-07   
9       2019-07-30      2020-10-04             2020-10-04   

  last_order_date_offline  order_num_total_ever_online  \
0              2021-02-26                          4.0   
1              2020-01-10                         19.0   
2              2019-12-01                          3.0   
3              2021-01-06                          1.0   
4              2019-08-03                          1.0   
5              2021-03-13                          1.0   
6              2020-03-04                          3.0   
7              2020-08-12                          1.0   
8              2020-01-25                          3.0   
9              2019-07-30                          1.0   

   order_num_total_ever_offline  customer_value_total_ever_offline  \
0                           1.0                             139.99   
1                           2.0                             159.97   
2                           2.0                             189.97   
3                           1.0                              39.99   
4                           1.0                              49.99   
5                           2.0                             150.87   
6                           1.0                              59.99   
7                           1.0                              49.99   
8                           2.0                             120.48   
9                           1.0                              69.98   

   customer_value_total_ever_online       interested_in_categories_12  
0                            799.38                           [KADIN]  
1                           1853.58  [ERKEK, COCUK, KADIN, AKTIFSPOR]  
2                            395.35                    [ERKEK, KADIN]  
3                             81.98               [AKTIFCOCUK, COCUK]  
4                            159.99                       [AKTIFSPOR]  
5                             49.99                           [KADIN]  
6                            315.94                       [AKTIFSPOR]  
7                            113.64                           [COCUK]  
8                            934.21             [ERKEK, COCUK, KADIN]  
9                             95.98                [KADIN, AKTIFSPOR]

In [3]:
df.describe().T



count        mean         std    min  \
order_num_total_ever_online        19945.0    3.110855    4.225647   1.00   
order_num_total_ever_offline       19945.0    1.913913    2.062880   1.00   
customer_value_total_ever_offline  19945.0  253.922597  301.532853  10.00   
customer_value_total_ever_online   19945.0  497.321690  832.601886  12.99   

                                      25%     50%     75%       max  
order_num_total_ever_online          1.00    2.00    4.00    200.00  
order_num_total_ever_offline         1.00    1.00    2.00    109.00  
customer_value_total_ever_offline   99.99  179.98  319.97  18119.14  
customer_value_total_ever_online   149.98  286.46  578.44  45220.13

In [4]:
def analyze_columns(df):
    """
    Veri çerçevesindeki değişkenlerin türlerini ve eksik değerlerini analiz eder.

    Args:
        df (DataFrame): Analiz edilecek veri çerçevesi.

    Returns:
        dict: Değişken türlerini ve eksik değerleri içeren bir sözlük.
    """
    # Kategorik sütunlar
    categorical_cols = df.select_dtypes(include=['object', 'category']).columns.tolist()

    # Numerik sütunlar
    numeric_cols = df.select_dtypes(include=['number']).columns.tolist()

    # Tarih sütunlar
    datetime_cols = df.select_dtypes(include=['datetime']).columns.tolist()

    # Boolean sütunlar
    bool_cols = df.select_dtypes(include=['bool']).columns.tolist()

    # Numerik görünen kategorik sütunlar
    numeric_but_categorical = [col for col in numeric_cols if df[col].nunique() < 10]

    # Kategorik görünen kardinal sütunlar
    cardinal_categoricals = [col for col in categorical_cols if df[col].nunique() > 30]

    # Eksik değer içeren sütunlar
    nan_cols = [col for col in df.columns if df[col].isna().any()]

    # Eksik değer içeren sütunların hangi gruplarda olduğu bilgisi
    col_analysis = {
        "categorical_cols": categorical_cols,
        "numeric_cols": numeric_cols,
        "datetime_cols": datetime_cols,
        "bool_cols": bool_cols,
        "numeric_but_categorical": numeric_but_categorical,
        "cardinal_categoricals": cardinal_categoricals
    }
    nan_cols_groups = {col: next(group for group, cols in col_analysis.items() if col in cols) for col in nan_cols}

    # Gözlem sayısı
    Observations = df.shape[0]

    # Değişken sayısı
    Variables = df.shape[1]

    print({
        "Observations": Observations,
        "Variables": Variables,
        "Categorical Columns": categorical_cols,
        "Numerical Columns": numeric_cols,
        "Datetime Columns": datetime_cols,
        "Boolean Columns": bool_cols,
        "Numerical but Categorical Columns": numeric_but_categorical,
        "Cardinal Categorical Columns": cardinal_categoricals,
        "Columns With NaN Values": nan_cols,
        "Variable Types of Columns with Nan Values": nan_cols_groups
    })

    return categorical_cols, numeric_cols, cardinal_categoricals, numeric_but_categorical

categorical_cols, numeric_cols, cardinal_categoricals, numeric_but_categorical = analyze_columns(df)

{'Observations': 19945, 'Variables': 12, 'Categorical Columns': ['master_id', 'order_channel', 'last_order_channel', 'first_order_date', 'last_order_date', 'last_order_date_online', 'last_order_date_offline', 'interested_in_categories_12'], 'Numerical Columns': ['order_num_total_ever_online', 'order_num_total_ever_offline', 'customer_value_total_ever_offline', 'customer_value_total_ever_online'], 'Datetime Columns': [], 'Boolean Columns': [], 'Numerical but Categorical Columns': [], 'Cardinal Categorical Columns': ['master_id', 'first_order_date', 'last_order_date', 'last_order_date_online', 'last_order_date_offline', 'interested_in_categories_12'], 'Columns With NaN Values': [], 'Variable Types of Columns with Nan Values': {}}


Omnichannel customers are defined as those who shop on both online and offline platforms.
Let's create new variables for each customer's total number of purchases and total spending.

In [5]:
df["order_num_total"] = df["order_num_total_ever_online"] + df["order_num_total_ever_offline"]
df["customer_value_total"] = df["customer_value_total_ever_offline"] + df["customer_value_total_ever_online"]


"date" variables seem as objects, so convert them to DateTime type.

In [6]:
df[[col for col in df.columns if 'date' in col]] = df[[col for col in df.columns if 'date' in col]].apply(pd.to_datetime)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19945 entries, 0 to 19944
Data columns (total 14 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   master_id                          19945 non-null  object        
 1   order_channel                      19945 non-null  object        
 2   last_order_channel                 19945 non-null  object        
 3   first_order_date                   19945 non-null  datetime64[ns]
 4   last_order_date                    19945 non-null  datetime64[ns]
 5   last_order_date_online             19945 non-null  datetime64[ns]
 6   last_order_date_offline            19945 non-null  datetime64[ns]
 7   order_num_total_ever_online        19945 non-null  float64       
 8   order_num_total_ever_offline       19945 non-null  float64       
 9   customer_value_total_ever_offline  19945 non-null  float64       
 10  customer_value_total_ever_online  

Analyze the distribution of the number of customers, total number of products purchased, and total spending across different shopping channels.

In [7]:
df.groupby("order_channel").agg({"master_id":"count",
                                 "order_num_total":"sum",
                                 "customer_value_total":"sum"})


master_id  order_num_total  customer_value_total
order_channel                                                  
Android App         9495          52269.0            7819062.76
Desktop             2735          10920.0            1610321.46
Ios App             2833          15351.0            2525999.93
Mobile              4882          21679.0            3028183.16

The top 10 customers with the highest spending.

In [8]:
df.sort_values("customer_value_total", ascending=False).head(10)

master_id order_channel last_order_channel  \
11150  5d1c466a-9cfd-11e9-9897-000d3a38a36f   Android App            Desktop   
4315   d5ef8058-a5c6-11e9-a2fc-000d3a38a36f   Android App        Android App   
7613   73fd19aa-9e37-11e9-9897-000d3a38a36f       Ios App            Offline   
13880  7137a5c0-7aad-11ea-8f20-000d3a38a36f       Ios App            Offline   
9055   47a642fe-975b-11eb-8c2a-000d3a38a36f   Android App            Offline   
7330   a4d534a2-5b1b-11eb-8dbd-000d3a38a36f       Desktop            Desktop   
8068   d696c654-2633-11ea-8e1c-000d3a38a36f       Ios App            Ios App   
163    fef57ffa-aae6-11e9-a2fc-000d3a38a36f        Mobile            Desktop   
7223   cba59206-9dd1-11e9-9897-000d3a38a36f   Android App        Android App   
18767  fc0ce7a4-9d87-11e9-9897-000d3a38a36f       Desktop            Desktop   

      first_order_date last_order_date last_order_date_online  \
11150       2013-10-11      2021-04-30             2021-04-30   
4315        2018-08-06      2021-02-23             2021-02-23   
7613        2014-01-14      2021-05-18             2021-01-30   
13880       2021-03-01      2021-04-13             2021-03-18   
9055        2021-04-07      2021-04-27             2021-04-07   
7330        2020-02-16      2021-04-30             2021-04-30   
8068        2017-05-10      2021-04-13             2021-04-13   
163         2016-11-08      2021-05-12             2021-05-12   
7223        2013-02-21      2021-05-09             2021-05-09   
18767       2018-11-24      2020-11-11             2020-11-11   

      last_order_date_offline  order_num_total_ever_online  \
11150              2020-12-24                        200.0   
4315               2020-07-06                         67.0   
7613               2021-05-18                         81.0   
13880              2021-04-13                         10.0   
9055               2021-04-27                          1.0   
7330               2020-12-18                         66.0   
8068               2019-08-15                         69.0   
163                2020-07-09                         36.0   
7223               2020-01-25                        130.0   
18767              2019-12-06                         18.0   

       order_num_total_ever_offline  customer_value_total_ever_offline  \
11150                           2.0                             684.97   
4315                            1.0                             130.49   
7613                            1.0                            1263.76   
13880                           1.0                             538.94   
9055                            3.0                           18119.14   
7330                            4.0                             843.68   
8068                            1.0                              82.48   
163                             1.0                             180.73   
7223                            1.0                              49.99   
18767                           2.0                              64.97   

       customer_value_total_ever_online  \
11150                          45220.13   
4315                           36687.80   
7613                           32654.34   
13880                          30688.47   
9055                            2587.20   
7330                           17599.89   
8068                           16836.09   
163                            12545.37   
7223                           12232.25   
18767                          12038.18   

                        interested_in_categories_12  order_num_total  \
11150  [AKTIFCOCUK, ERKEK, COCUK, KADIN, AKTIFSPOR]            202.0   
4315          [AKTIFCOCUK, ERKEK, KADIN, AKTIFSPOR]             68.0   
7613               [ERKEK, COCUK, KADIN, AKTIFSPOR]             82.0   
13880                     [ERKEK, KADIN, AKTIFSPOR]             11.0   
9055                                    [AKTIFSPOR]              4.0   
7330                      [ERKEK,

The top 10 customers with the highest order.

In [9]:
df.sort_values("order_num_total",ascending=False).head(10)

master_id order_channel last_order_channel  \
11150  5d1c466a-9cfd-11e9-9897-000d3a38a36f   Android App            Desktop   
7223   cba59206-9dd1-11e9-9897-000d3a38a36f   Android App        Android App   
8783   a57f4302-b1a8-11e9-89fa-000d3a38a36f   Android App            Offline   
2619   fdbe8304-a7ab-11e9-a2fc-000d3a38a36f   Android App            Offline   
6322   329968c6-a0e2-11e9-a2fc-000d3a38a36f       Ios App            Ios App   
7613   73fd19aa-9e37-11e9-9897-000d3a38a36f       Ios App            Offline   
9347   44d032ee-a0d4-11e9-a2fc-000d3a38a36f        Mobile             Mobile   
10954  b27e241a-a901-11e9-a2fc-000d3a38a36f        Mobile             Mobile   
8068   d696c654-2633-11ea-8e1c-000d3a38a36f       Ios App            Ios App   
7330   a4d534a2-5b1b-11eb-8dbd-000d3a38a36f       Desktop            Desktop   

      first_order_date last_order_date last_order_date_online  \
11150       2013-10-11      2021-04-30             2021-04-30   
7223        2013-02-21      2021-05-09             2021-05-09   
8783        2019-08-07      2020-11-04             2020-09-07   
2619        2018-10-18      2020-06-30             2018-10-18   
6322        2019-02-14      2021-04-05             2021-04-05   
7613        2014-01-14      2021-05-18             2021-01-30   
9347        2019-02-11      2021-02-11             2021-02-11   
10954       2015-09-12      2021-04-01             2021-04-01   
8068        2017-05-10      2021-04-13             2021-04-13   
7330        2020-02-16      2021-04-30             2021-04-30   

      last_order_date_offline  order_num_total_ever_online  \
11150              2020-12-24                        200.0   
7223               2020-01-25                        130.0   
8783               2020-11-04                          2.0   
2619               2020-06-30                          1.0   
6322               2020-02-17                          2.0   
7613               2021-05-18                         81.0   
9347               2020-12-24                          3.0   
10954              2019-10-08                         72.0   
8068               2019-08-15                         69.0   
7330               2020-12-18                         66.0   

       order_num_total_ever_offline  customer_value_total_ever_offline  \
11150                           2.0                             684.97   
7223                            1.0                              49.99   
8783                          109.0                           10239.46   
2619                           87.0                            8432.25   
6322                           81.0                            3997.55   
7613                            1.0                            1263.76   
9347                           74.0                            4640.77   
10954                           3.0                             292.93   
8068                            1.0                              82.48   
7330                            4.0                             843.68   

       customer_value_total_ever_online  \
11150                          45220.13   
7223                           12232.25   
8783                             143.98   
2619                             139.98   
6322                             242.81   
7613                           32654.34   
9347                             543.28   
10954                           5004.95   
8068                           16836.09   
7330                           17599.89   

                        interested_in_categories_12  order_num_total  \
11150  [AKTIFCOCUK, ERKEK, COCUK, KADIN, AKTIFSPOR]            202.0   
7223          [AKTIFCOCUK, ERKEK, KADIN, AKTIFSPOR]            131.0   
8783   [AKTIFCOCUK, ERKEK, COCUK, KADIN, AKTIFSPOR]            111.0   
2619                                        [ERKEK]             88.0   
6322                                        [ERKEK]             83.0   
7613               [ERKEK, COCUK,

It may be wanted to be this process functionalized in order to use later. This part will be added as a bonus.

In [10]:
def preprocess_data(df):
    """
    Veri setini ön işleyerek gerekli dönüşümleri yapar.

    Parameters:
    df (pd.DataFrame): İşlenecek DataFrame.

    Returns:
    pd.DataFrame: İşlenmiş DataFrame.
    """
    # Toplam sipariş sayısını ve toplam müşteri değerini hesapla
    df["order_num_total"] = df["order_num_total_ever_online"] + df["order_num_total_ever_offline"]
    df["customer_value_total"] = df["customer_value_total_ever_offline"] + df["customer_value_total_ever_online"]

    df[[col for col in df.columns if 'date' in col]] = df[[col for col in df.columns if 'date' in col]].apply(pd.to_datetime)

    return df

### **1. Customer Segmentation with RFM Analysis**
RFM stands for Recency, Frequency, and Monetary.

Recency: How many days have passed since the customer's last order?
Frequency: The number of times the customer has placed an order. In other words, how many orders has the customer made?
Monetary: The total amount of money the customer has spent on their orders.
These metrics will be extracted from the data using feature engineering, and the analysis will be performed based on these values.

Why do we need these metrics?
We have various values in our dataset, and the challenge is how to segment these values meaningfully. For instance, consider the total amount spent by customers (monetary value) and examine the results. Using RFM analysis, we can effectively segment customers based on their purchasing behaviors, allowing for more targeted marketing strategies and better customer relationship management.

In [11]:
df["customer_value_total"].describe()

count    19945.000000
mean       751.244287
std        895.402173
min         44.980000
25%        339.980000
50%        545.270000
75%        897.780000
max      45905.100000
Name: customer_value_total, dtype: float64

The minimum value is 44.980, and the maximum value is 45905.100. Even the quantiles are variable. In order to scale them, the metrics are used, and the points are given according to these values. The highest values for monetary and frequency, and the lowest values for frequency(since there are new customers) will take the highest scores. In this project, scores will be taken from values between 1 and 5, from low to high.

### Business Problem:
FLO, an online shoe store, wants to segment its customers and determine marketing strategies according to these segments. To this end, the behaviors of the customers will be defined and groups will be formed according to the clustering of these behaviors.

#### Choose Analyze Date
Since the dataset consists of 2020-2021 data, take analyze date as 2 days later after the last date in the dataset. Today's date is not taken because the day difference with analysis and data in the dataset will be considered, for prevent the confusion of metrics, 2 days later will be taken. It seems a logical approach since these kinds of datasets are kept up-to-date periodically, and analyses of them are conducted with up-to-date datasets. Therefore, our approach is to consider taking the dataset like it's up-to-date.

In [12]:
df["last_order_date"].max()  # 2021-05-30
analysis_date = dt.datetime(2021, 6, 1)

##### RFM Metrics
Since these metrics are needed and analyze will be done according the these metric values, create a new dataset from the main dataset.

In [13]:
df_rfm = pd.DataFrame()
df_rfm["customer_id"] = df["master_id"]
df_rfm["recency"] = (analysis_date - df["last_order_date"]).dt.days
# df_rfm["recency"] = (analysis_date - df["last_order_date"]).astype('timedelta64[D]')
df_rfm["frequency"] = df["order_num_total"]
df_rfm["monetary"] = df["customer_value_total"]
df_rfm.head()

customer_id  recency  frequency  monetary
0  cc294636-19f0-11eb-8d74-000d3a38a36f       95        5.0    939.37
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f      105       21.0   2013.55
2  69b69676-1a40-11ea-941b-000d3a38a36f      186        5.0    585.32
3  1854e56c-491f-11eb-806e-000d3a38a36f      135        2.0    121.97
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       86        2.0    209.98

In [14]:
df_rfm["recency_score"] = pd.qcut(df_rfm['recency'], 5, labels=[5, 4, 3, 2, 1])
df_rfm["frequency_score"] = pd.qcut(df_rfm['frequency'].rank(method="first"), 5, labels=[1, 2, 3, 4, 5])
df_rfm["monetary_score"] = pd.qcut(df_rfm['monetary'], 5, labels=[1, 2, 3, 4, 5])
df_rfm.head()

customer_id  recency  frequency  monetary  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f       95        5.0    939.37   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f      105       21.0   2013.55   
2  69b69676-1a40-11ea-941b-000d3a38a36f      186        5.0    585.32   
3  1854e56c-491f-11eb-806e-000d3a38a36f      135        2.0    121.97   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       86        2.0    209.98   

  recency_score frequency_score monetary_score  
0             3               4              4  
1             3               5              5  
2             2               4              3  
3             3               1              1  
4             3               1              1

Combine recency_score and frequency_score into a single variable and save it as RF_SCORE.

In [15]:
df_rfm["RF_SCORE"] = (df_rfm['recency_score'].astype(str) + df_rfm['frequency_score'].astype(str))
df_rfm.head()

customer_id  recency  frequency  monetary  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f       95        5.0    939.37   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f      105       21.0   2013.55   
2  69b69676-1a40-11ea-941b-000d3a38a36f      186        5.0    585.32   
3  1854e56c-491f-11eb-806e-000d3a38a36f      135        2.0    121.97   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       86        2.0    209.98   

  recency_score frequency_score monetary_score RF_SCORE  
0             3               4              4       34  
1             3               5              5       35  
2             2               4              3       24  
3             3               1              1       31  
4             3               1              1       31

##### Segmentation of RF Scores
Looking for customers that are loyal to the company. Therefore, recency and frequency are crucial in this analysis. Monetary value seems as considerable, but here loyalty surpasses the money that the customer brings, so we are not interested in a customer that brings a high amount of money but doesn't return back. The analysis is based on the customers that shop periodically. Hence, the RF score will be used rather than the RFM score on segmentation. How the segmentation will be specified is shown in the below figure:

![alt text](cltv_grafik.png)

In [16]:
seg_map = {
    r'[1-2][1-2]': 'hibernating',
    r'[1-2][3-4]': 'at_Risk',
    r'[1-2]5': 'cant_loose',
    r'3[1-2]': 'about_to_sleep',
    r'33': 'need_attention',
    r'[3-4][4-5]': 'loyal_customers',
    r'41': 'promising',
    r'51': 'new_customers',
    r'[4-5][2-3]': 'potential_loyalists',
    r'5[4-5]': 'champions'
}
df_rfm['segment'] = df_rfm['RF_SCORE'].replace(seg_map, regex=True)
df_rfm.head()

customer_id  recency  frequency  monetary  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f       95        5.0    939.37   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f      105       21.0   2013.55   
2  69b69676-1a40-11ea-941b-000d3a38a36f      186        5.0    585.32   
3  1854e56c-491f-11eb-806e-000d3a38a36f      135        2.0    121.97   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f       86        2.0    209.98   

  recency_score frequency_score monetary_score RF_SCORE          segment  
0             3               4              4       34  loyal_customers  
1             3               5              5       35  loyal_customers  
2             2               4              3       24          at_Risk  
3             3               1              1       31   about_to_sleep  
4             3               1              1       31   about_to_sleep

In [17]:
df_rfm[["segment", "recency", "frequency", "monetary"]].groupby("segment").agg(["mean", "count"])

recency        frequency           monetary      
                           mean count       mean count         mean count
segment                                                                  
about_to_sleep       113.785144  1629   2.401473  1629   359.008963  1629
at_Risk              241.606835  3131   4.472373  3131   646.610236  3131
cant_loose           235.444167  1200  10.698333  1200  1474.468217  1200
champions             17.106625  1932   8.934265  1932  1406.625109  1932
hibernating          247.949501  3604   2.394007  3604   366.267062  3604
loyal_customers       82.594763  3361   8.374591  3361  1216.818616  3361
need_attention       113.828676   823   3.727825   823   562.142965   823
new_customers         17.917647   680   2.000000   680   339.955544   680
potential_loyalists   37.155888  2938   3.304289  2938   533.184466  2938
promising             58.921175   647   2.000000   647   335.672705   647

##### Example Cases for Usage of These Segments
##### Case Example 1:
FLO is incorporating a new women's shoe brand. The product prices of the included brand are above the general customer preferences. For this reason, it is desired to contact the customers in the profile that will be interested in the promotion of the brand and product sales. Let's find the IDs of the customers who will be contacted specifically, those who shop from their loyal customers (champions, loyal_customers) and women category.

In [18]:
hedef_müsteri_ids = df_rfm[df_rfm["segment"].isin(["Champions", "Loyal_customers"])]["customer_id"]

kadın_kategorisi_alısveris = df.loc[(df["master_id"].isin(hedef_müsteri_ids)) & (df["interested_in_categories_12"].str.contains("KADIN")), "master_id"]

kadın_kategorisi_alısveris.head(10)



Series([], Name: master_id, dtype: object)

##### Case Example 2
Nearly 40% discount is planned for Men's and Children's products. It is aimed to specifically target customers who are good customers in the past, but who have not shopped for a long time, who are interested in categories related to this discount, who should not be lost, those who are asleep and new customers. Find the IDs of the customers in the appropriate profile.

In [19]:
hedef_musteri_ids2 = df_rfm[df_rfm["segment"].isin(["Cant_loose", "Hibernating", "New_customers"])]["customer_id"]
erkek_cocuk_alısveris = df[df["master_id"].isin(hedef_musteri_ids2) & df["interested_in_categories_12"].str.contains("ERKEK|COCUK")]["master_id"]

erkek_cocuk_alısveris.head(10)

Series([], Name: master_id, dtype: object)

### **2. CLTV Prediction with Gamma-Gamma and BD-NBD Models**
Firstly, it has to be understood why this calculation is conducted, and the prediction is made. The reason for this is planning. It means that companies not only have to segment the customers, but also they have to predict how many customers they will have and what the future profit can be. At that point, Customer Life Time Value(CLTV) calculations and predictions are made for answering these questions.

*Calculation of CLTV*

**CLTV** = (Customer Value / Churn Rate) * Profit Margin

**Customer Value** = Average Order Value * Purchase Frequency
Avarage Order Value = Total Price / Total Transaction
Purchase Frequency = Total Transaction / Total Number of Customers

**Churn Rate** = 1 - Repeat Rate
Repeat Rate = The Number of Customers Giving Multiple Order / Total Number of Customers**

**Profit Margin** = Total Price * Desired Profit

#### Prediction

**CLTV** = (Customer Value / Churn Rate) * Profit Margin

**Customer Value** = Purchase Frequency * Average Order Value

**CLTV** = Expected Number of Transaction * Expected Average Profit*

**CLTV Prediction = BGNBD Model * Gamma Gamma Model**

Business Problem:
FLO wants to set a roadmap for sales and marketing activities. In order for the company to make a medium-long-term plan, it is necessary to estimate the potential value that existing customers will provide to the company in the future.

Prepare Dataset
The same dataset will be used, but there will be different processes will be made. Obviously, there are outliers in the dataset that is mentioned on the RFM part. On CLTV analysis, these values will not be used directly, firstly, these values will be suppressed, which means the value will be changed. Also, when CLTV is calculated, frequency values have to be integers, therefore, these will be rounded.

In [20]:
import pandas as pd
import datetime as dt
from lifetimes import BetaGeoFitter
from lifetimes import GammaGammaFitter

In [21]:
df = pd.read_csv("C:\\Users\\utku\\Desktop\\miuul\\FLOCLTVPrediction\\flo_data_20k.csv")


In [22]:
def detect_and_replace_outliers(dataframe):
    """
    Verilen DataFrame'deki outlier'ları tespit eder ve baskılar.

    Parameters:
    dataframe (pd.DataFrame): İşlenecek DataFrame.
    threshold (float): outlier olarak kabul edilecek standart sapma eşiği. Varsayılan değer 1.5.

    Returns:
    pd.DataFrame: Aykırı değerlerin baskılandığı DataFrame.
    """
    outlier_columns = []

    for col in dataframe.columns:
        if pd.api.types.is_numeric_dtype(dataframe[col]):
            Q1 = dataframe[col].quantile(0.01)
            Q3 = dataframe[col].quantile(0.99)
            IQR = Q3 - Q1
            lower_bound = Q1 - 1.5 * IQR
            upper_bound = Q3 + 1.5 * IQR

            outliers = dataframe[(dataframe[col] < lower_bound) | (dataframe[col] > upper_bound)]
            if not outliers.empty:
                outlier_columns.append(col)

            dataframe.loc[dataframe[col] < lower_bound, col] = round(lower_bound, 0)
            dataframe.loc[dataframe[col] > upper_bound, col] = round(upper_bound, 0)
    print("Columns with Outliers:", outlier_columns)

    return dataframe

In [23]:
#First check values' min,max and std 
df.describe()

order_num_total_ever_online  order_num_total_ever_offline  \
count                 19945.000000                  19945.000000   
mean                      3.110855                      1.913913   
std                       4.225647                      2.062880   
min                       1.000000                      1.000000   
25%                       1.000000                      1.000000   
50%                       2.000000                      1.000000   
75%                       4.000000                      2.000000   
max                     200.000000                    109.000000   

       customer_value_total_ever_offline  customer_value_total_ever_online  
count                       19945.000000                      19945.000000  
mean                          253.922597                        497.321690  
std                           301.532853                        832.601886  
min                            10.000000                         12.990000  
25%                            99.990000                        149.980000  
50%                           179.980000                        286.460000  
75%                           319.970000                        578.440000  
max                         18119.140000                      45220.130000

In [24]:
detect_and_replace_outliers(df)

Columns with Outliers: ['order_num_total_ever_online', 'order_num_total_ever_offline', 'customer_value_total_ever_offline', 'customer_value_total_ever_online']


master_id order_channel last_order_channel  \
0      cc294636-19f0-11eb-8d74-000d3a38a36f   Android App            Offline   
1      f431bd5a-ab7b-11e9-a2fc-000d3a38a36f   Android App             Mobile   
2      69b69676-1a40-11ea-941b-000d3a38a36f   Android App        Android App   
3      1854e56c-491f-11eb-806e-000d3a38a36f   Android App        Android App   
4      d6ea1074-f1f5-11e9-9346-000d3a38a36f       Desktop            Desktop   
...                                     ...           ...                ...   
19940  727e2b6e-ddd4-11e9-a848-000d3a38a36f   Android App            Offline   
19941  25cd53d4-61bf-11ea-8dd8-000d3a38a36f       Desktop            Desktop   
19942  8aea4c2a-d6fc-11e9-93bc-000d3a38a36f       Ios App            Ios App   
19943  e50bb46c-ff30-11e9-a5e8-000d3a38a36f   Android App        Android App   
19944  740998d2-b1f7-11e9-89fa-000d3a38a36f   Android App        Android App   

      first_order_date last_order_date last_order_date_online  \
0           2020-10-30      2021-02-26             2021-02-21   
1           2017-02-08      2021-02-16             2021-02-16   
2           2019-11-27      2020-11-27             2020-11-27   
3           2021-01-06      2021-01-17             2021-01-17   
4           2019-08-03      2021-03-07             2021-03-07   
...                ...             ...                    ...   
19940       2019-09-21      2020-07-05             2020-06-05   
19941       2020-03-01      2020-12-22             2020-12-22   
19942       2019-09-11      2021-05-24             2021-05-24   
19943       2019-03-27      2021-02-13             2021-02-13   
19944       2019-09-03      2020-06-06             2020-06-06   

      last_order_date_offline  order_num_total_ever_online  \
0                  2021-02-26                          4.0   
1                  2020-01-10                         19.0   
2                  2019-12-01                          3.0   
3                  2021-01-06                          1.0   
4                  2019-08-03                          1.0   
...                       ...                          ...   
19940              2020-07-05                          1.0   
19941              2020-03-01                          1.0   
19942              2019-09-11                          2.0   
19943              2021-01-08                          1.0   
19944              2019-09-03                          1.0   

       order_num_total_ever_offline  customer_value_total_ever_offline  \
0                               1.0                             139.99   
1                               2.0                             159.97   
2                               2.0                             189.97   
3                               1.0                              39.99   
4                               1.0                              49.99   
...                             ...                                ...   
19940                           2.0                             289.98   
19941                           1.0                             150.48   
19942                           1.0                             139.98   
19943                           5.0                             711.79   
19944                           1.0                              39.99   

       customer_value_total_ever_online       interested_in_categories_12  
0                                799.38                           [KADIN]  
1                               1853.58  [ERKEK, COCUK, KADIN, AKTIFSPOR]  
2                                395.35                    [ERKEK, KADIN]  
3                                 81.98               [AKTIFCOCUK, COCUK]  
4                                159.99                       [AKTIFSPOR]  
...                                 ...                               ...  
19940                            111.98                [ERKEK, AKTIFSPOR]  
19941                            239.99           

In [26]:
#Check again after capping outliers
df.describe()

order_num_total_ever_online  order_num_total_ever_offline  \
count                 19945.000000                  19945.000000   
mean                      3.091953                      1.886187   
std                       3.809541                      1.434694   
min                       1.000000                      1.000000   
25%                       1.000000                      1.000000   
50%                       2.000000                      1.000000   
75%                       4.000000                      2.000000   
max                      48.000000                     16.000000   

       customer_value_total_ever_offline  customer_value_total_ever_online  
count                        19945.00000                      19945.000000  
mean                           251.92132                        489.705676  
std                            251.02367                        632.609844  
min                             10.00000                         12.990000  
25%                             99.99000                        149.980000  
50%                            179.98000                        286.460000  
75%                            319.97000                        578.440000  
max                           3020.00000                       7800.000000

In [27]:
#Creating new variable for each customer's total order and spending 

df["order_num_total"] = df["order_num_total_ever_online"] + df["order_num_total_ever_offline"]
df["customer_value_total"] = df["customer_value_total_ever_offline"] + df["customer_value_total_ever_online"]

In [28]:
#Convert the variables include "data" but seem as objects convert to DateTime type.
df[[col for col in df.columns if 'date' in col]] = df[[col for col in df.columns if 'date' in col]].apply(pd.to_datetime)

#### **Creating the CLTV Data Structure**

Taking the date of the last purchase in the dataset and consider it as the analysis date, two days after the purchase date

In [ ]:
df["last_order_date"].max()
analysis_date = dt.datetime(2021, 6, 1)

In [29]:
cltv_df = pd.DataFrame()
cltv_df["customer_id"] = df["master_id"]
cltv_df["recency_cltv_weekly"] = ((df["last_order_date"] - df["first_order_date"]).dt.days) / 7
cltv_df["T_weekly"] = ((analysis_date - df["first_order_date"]).astype('timedelta64[D]')) / 7
cltv_df["frequency"] = df["order_num_total"]
cltv_df["monetary_cltv_avg"] = df["customer_value_total"] / df["order_num_total"]

**1. BG/NBD model**

In [30]:
bgf = BetaGeoFitter(penalizer_coef=0.001)
bgf.fit(cltv_df['frequency'],
        cltv_df['recency_cltv_weekly'],
        cltv_df['T_weekly'])

c:\Users\utku\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\arraylike.py:402: RuntimeWarning: invalid value encountered in sqrt
  result = getattr(ufunc, method)(*inputs, **kwargs)


<lifetimes.BetaGeoFitter: fitted with 19945 subjects, a: 0.00, alpha: 76.17, b: 0.00, r: 3.66>

Predicted purchases expected from customers within 3 months

In [31]:
cltv_df["exp_sales_3_month"] = bgf.predict(4 * 3,
                                           cltv_df['frequency'],
                                           cltv_df['recency_cltv_weekly'],
                                           cltv_df['T_weekly'])

Predicted purchases expected from customers within 6 months

In [32]:
cltv_df["exp_sales_6_month"] = bgf.predict(4 * 6,
                                           cltv_df['frequency'],
                                           cltv_df['recency_cltv_weekly'],
                                           cltv_df['T_weekly'])

The top 10 customers who will make the most purchases in the 3rd and 6th months.

In [33]:
cltv_df.sort_values("exp_sales_3_month", ascending=False).head(10)

customer_id  recency_cltv_weekly    T_weekly  \
7330   a4d534a2-5b1b-11eb-8dbd-000d3a38a36f            62.714286   67.285714   
15611  4a7e875e-e6ce-11ea-8f44-000d3a38a36f            39.714286   40.000000   
8328   1902bf80-0035-11eb-8341-000d3a38a36f            28.857143   33.285714   
19538  55d54d9e-8ac7-11ea-8ec0-000d3a38a36f            52.571429   58.714286   
14373  f00ad516-c4f4-11ea-98f7-000d3a38a36f            38.000000   46.428571   
10489  7af5cd16-b100-11e9-9757-000d3a38a36f           103.142857  111.857143   
4315   d5ef8058-a5c6-11e9-a2fc-000d3a38a36f           133.142857  147.142857   
6756   27310582-6362-11ea-a6dc-000d3a38a36f            62.714286   64.142857   
6666   53fe00d4-7b7a-11eb-960b-000d3a38a36f             9.714286   13.000000   
10536  e143b6fa-d6f8-11e9-93bc-000d3a38a36f           104.571429  113.428571   

       frequency  monetary_cltv_avg  exp_sales_3_month  exp_sales_6_month  
7330        52.0         166.224615           4.656138           9.312276  
15611       29.0         165.297586           3.373958           6.747915  
8328        25.0          97.439600           3.142396           6.284792  
19538       31.0         228.530000           3.083779           6.167558  
14373       27.0         141.354815           3.001287           6.002574  
10489       43.0         157.112558           2.978047           5.956093  
4315        49.0         161.846735           2.829904           5.659808  
6756        29.0         168.881034           2.793429           5.586858  
6666        17.0         259.865294           2.780689           5.561378  
10536       40.0         176.200000           2.763492           5.526983

In [34]:
#Fitting the Gamma-Gamma model to predict the average values that customers will generate.

ggf = GammaGammaFitter(penalizer_coef=0.01)
ggf.fit(cltv_df['frequency'], cltv_df['monetary_cltv_avg'])
cltv_df["exp_average_value"] = ggf.conditional_expected_average_profit(cltv_df['frequency'],
                                                                       cltv_df['monetary_cltv_avg'])

Calculate the CLTV for 3 and 6 months and add it to the dataframe with the name "cltv"

In [35]:
cltv = ggf.customer_lifetime_value(bgf,
                                   cltv_df['frequency'],
                                   cltv_df['recency_cltv_weekly'],
                                   cltv_df['T_weekly'],
                                   cltv_df['monetary_cltv_avg'],
                                   time=6,
                                   freq="W",
                                   discount_rate=0.01)

In [36]:
cltv_df["cltv"] = cltv
cltv_df.head()

customer_id  recency_cltv_weekly    T_weekly  \
0  cc294636-19f0-11eb-8d74-000d3a38a36f            17.000000   30.571429   
1  f431bd5a-ab7b-11e9-a2fc-000d3a38a36f           209.857143  224.857143   
2  69b69676-1a40-11ea-941b-000d3a38a36f            52.285714   78.857143   
3  1854e56c-491f-11eb-806e-000d3a38a36f             1.571429   20.857143   
4  d6ea1074-f1f5-11e9-9346-000d3a38a36f            83.142857   95.428571   

   frequency  monetary_cltv_avg  exp_sales_3_month  exp_sales_6_month  \
0        5.0         187.874000           0.973927           1.947853   
1       21.0          95.883333           0.983161           1.966323   
2        5.0         117.064000           0.670586           1.341172   
3        2.0          60.985000           0.700412           1.400824   
4        2.0         104.990000           0.396039           0.792077   

   exp_average_value        cltv  
0         193.632679  395.733234  
1          96.665048  199.430693  
2         120.967619  170.224184  
3          67.320145   98.945526  
4         114.325108   95.011659

In [38]:
cltv_df.sort_values("cltv",ascending=False).head(10)

customer_id  recency_cltv_weekly    T_weekly  \
9055   47a642fe-975b-11eb-8c2a-000d3a38a36f             2.857143    7.857143   
13880  7137a5c0-7aad-11ea-8f20-000d3a38a36f             6.142857   13.142857   
17323  f59053e2-a503-11e9-a2fc-000d3a38a36f            51.714286  101.000000   
12438  625f40a2-5bd2-11ea-98b0-000d3a38a36f            74.285714   74.571429   
7330   a4d534a2-5b1b-11eb-8dbd-000d3a38a36f            62.714286   67.285714   
8868   9ce6e520-89b0-11ea-a6e7-000d3a38a36f             3.428571   34.428571   
6402   851de3b4-8f0c-11eb-8cb8-000d3a38a36f             8.285714    9.428571   
6666   53fe00d4-7b7a-11eb-960b-000d3a38a36f             9.714286   13.000000   
19538  55d54d9e-8ac7-11ea-8ec0-000d3a38a36f            52.571429   58.714286   
14858  031b2954-6d28-11eb-99c4-000d3a38a36f            14.857143   15.571429   

       frequency  monetary_cltv_avg  exp_sales_3_month  exp_sales_6_month  \
9055         4.0        1401.800000           1.094385           2.188769   
13880       11.0         758.085455           1.970108           3.940216   
17323        7.0        1106.467143           0.722238           1.444476   
12438       16.0         501.873750           1.565309           3.130618   
7330        52.0         166.224615           4.656138           9.312276   
8868         8.0         601.226250           1.265456           2.530912   
6402         2.0         862.690000           0.793924           1.587847   
6666        17.0         259.865294           2.780689           5.561378   
19538       31.0         228.530000           3.083779           6.167558   
14858        3.0         743.586667           0.871564           1.743128   

       exp_average_value         cltv  
9055         1449.060468  3327.777040  
13880         767.360602  3172.394390  
17323        1127.611525  1708.982063  
12438         506.166665  1662.613492  
7330          166.712253  1628.887381  
8868          611.492616  1623.812684  
6402          923.679965  1538.855906  
6666          262.072907  1529.227995  
19538         229.606946  1485.819156  
14858         778.050370  1422.999674

### **Creating Segments Based on CLTV**

In [42]:
cltv_df["cltv_segment"] = pd.qcut(cltv_df["cltv"], 4, labels=["D", "C", "B", "A"])

Reviewing the averages of recency, frequency, and monetary for segments

In [43]:
cltv_df.groupby("cltv_segment")["recency_cltv_weekly","frequency","monetary_cltv_avg"].mean()

recency_cltv_weekly  frequency  monetary_cltv_avg
cltv_segment                                                   
D                      139.000172   3.768799          93.151603
C                       92.629534   4.404733         125.788537
B                       81.988367   5.092659         160.636574
A                       67.427139   6.646611         228.831142